### PR TITLE
:recycle: Generalize pipeline ingestion

### PR DESCRIPTION
**Pull request by bpenteado**
- `IngestBfxRun` is now `IngestPipelineRun`, which provides compatibility with any kind of pipeline (bfx, ml, etc.).

### CODE REVIEW

File: lamindb/db/_ingest.py [hunk 1]
```diff
@@ -382,54 +360,77 @@ def _register_dobjects(self, jupynb_id, jupynb_v, jupynb_name):
...
                     f"{settings.user.handle} ({settings.user.id})",
                 ]
             )
 
-    def _register_samples(self):
-        """Register entries in the techsample, biosample, biometa tables."""
-        sample_id = self._run.sample_id
-        if sample_id is None:
-            # insert techsample, biosample, and biometa entries
+    def _ingest_samples(self):
```
Comment (bpenteado) [inline lamindb/db/_ingest.py:371]: @sunnyosun I would review the logic here carefully to make sure I correctly understood what we discussed.

File: lamindb/db/_ingest.py [hunk 2]
```diff
@@ -335,37 +334,16 @@ def log(self):
...
-    def _link_dobjects_bfxmeta(self):
-        """Register dobject's bfxmeta entry and link it to the dobject."""
-        for filepath, (dobject_id, _) in self.dobjects.items():
-            bfxmeta_id = self._register_bfxmeta(filepath)
-            insert.dobject_bfxmeta(dobject_id=dobject_id, bfxmeta_id=bfxmeta_id)
-
-    def _register(self, schema_module, pk: dict, **kwargs):
-        """Generic registration helper function."""
-        result = getattr(query, schema_module)(**pk).first()
+    def _ingest(self, table: str, pk: dict = {}, fk: dict = {}, **kwargs):
```
[the remark this answers is not in the document]
Comment (falexwolf) [reply lamindb/db/_ingest.py:337]: Sounds good, Benardo!

**Comment by bpenteado** [comment]
I also think this can be further simplified by making `IngestObject` and `IngestPipelineRun` inherit from a `IngestEntity` class, but decided to leave this for a further iteration so @falexwolf can come in quickly and implement non-atomic erroring.

**Comment by falexwolf** [comment]
Thanks, Bernardo!

If you have clarity on inhering from IngestObject, then you could also do that. It sounds a lot like it would help me do the erroring thing. 😅 I got a bit out of the loop with the whole code around ingest, so any simplification is very much welcome on my side!

And if that happens tomorrow, that's also OK! I'm trying to ship a pilot result today, but I guess they'll likely not play with it until tomorrow or early next week.

**Comment by bpenteado** [comment]
@sunnyosun PR is ready for your review! I’ll kindly ask you to pay close attention to the logic in `IngestPipelineRun._ingest_samples()` (permalink [here](https://github.com/laminlabs/lamindb/blob/899dba20765e497d808d6adf3be0fd1b101344d1/lamindb/db/_ingest.py#L371-L408)) and `IngestPipelineRun._ingest()` (permalink [here](https://github.com/laminlabs/lamindb/blob/899dba20765e497d808d6adf3be0fd1b101344d1/lamindb/db/_ingest.py#L337-L344)) to make sure that I understood the sample ingestion logic correctly and that I’m handling force insertion adequately.

**Comment by bpenteado** [comment]
@falexwolf I prefer not to implement inheritance for `IngestPipelineRun` and `IngestObject` here (I think it would be too much for a single PR). This PR already has a lot of complexity (new ingestion logic for samples + full refactoring). So let’s get this one through the door and I’ll make another PR right after? We can also leave the inheritance PR for later if you prefer to start working on non-atomic erroring right away.

**Comment by falexwolf** [comment]
Makes sense, Bernardo! It’d be great if you could continue with the inheritance one right after! I have an in-person meeting for the next couple of hours with a last potential angel (or potential advisor/mentor).

**Review by falexwolf — APPROVED** [review]
I think it looks great! Thank you! 🙏

**Comment by bpenteado** [comment]
Perfect! Should I go ahead with the merge or wait for Sunny’s review as well?

**Comment by falexwolf** [comment]
I think it’s perfectly OK to go ahead if you’d continue with the next PR immediately! Should there be anything from her side, one can still work it in!

**Comment by falexwolf** [comment]
In case of doubt, let’s always keep moving forward at pace and correct errors afterward rather than trying to get everything perfectly right the first time and losing time because of waiting for reviews. :sweat_smile:

[Of course, this doesn’t hold for production releases. But the main lamindb repo branch isn’t production-ready anyway right now.]

**Pull request merged**